### PR TITLE
Update field-offset and enable unstable_offset_of

### PR DESCRIPTION
**Pull request by est31**
This makes the compiler use the builtin `offset_of!()` macro, through the wrappers in memoffset and then in field-offset.

cc  #111839

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -3988,7 +3998,7 @@ version = "0.0.0"
 dependencies = [
  "field-offset",
  "measureme",
- "memoffset",
+ "memoffset 0.8.0",
```
Comment (klensy) [inline Cargo.lock:4001]: Bump here too? So only `crossbeam-epoch` uses memoffset 0.8 in workspace now.

File: compiler/rustc_middle/Cargo.toml [hunk 2]
```diff
@@ -40,5 +40,8 @@ smallvec = { version = "1.8.1", features = ["union", "may_dangle"] }
 thin-vec = "0.2.12"
 tracing = "0.1"
 
+# Not used directly, but included to enable the unstable_offset_of feature
+memoffset = { version = "0.9", features = ["unstable_offset_of"] }
+
```
Comment (klensy) [inline compiler/rustc_middle/Cargo.toml:45]: Can this be moved to `rustc_query_impl` where `field-offset` and `memoffset` used?

File: compiler/rustc_query_impl/Cargo.toml [hunk 3]
```diff
@@ -25,5 +24,8 @@ rustc_span = { path = "../rustc_span" }
 thin-vec = "0.2.12"
 tracing = "0.1"
 
+# Not used directly, but included to enable the unstable_offset_of feature
```
Comment (Nilstrieb) [inline compiler/rustc_query_impl/Cargo.toml:27]: is the comment still accurate? this crate had the dep before

Comment (est31) [reply compiler/rustc_query_impl/Cargo.toml:27]: Grep for `memoffset`, there is no results.

Before it was included to enable the `unstable_const` feature, which [<enabled usage of the `pointer::offset_from` function](https://github.com/Gilnaa/memoffset/blob/6afc606b09d293ae452c29f82c72cb0362bd6190/src/offset_of.rs#L48-L69), but that function is not used if `unstable_offset_of` is enabled as that directly uses `offset_of!()`.

The `offset_of_tuple` macro still doesn't use `offset_of` so `unstable_const` does *something* to the public API of `memoffset` still, but that macro is not used I think neither by rustc nor by any of its deps, at least that's my impression.

In other words, enabling `unstable_const` makes no sense at this point.
